[LoRA][III] Add LoRA support for MoE layers and enable TP

jonahbernard opened the pull request:
## Motivation

This PR adds support for LoRA serving on the expert layers for Mixture-of-Expert models. This is the initial development PR, which was split into two other smaller PRs, for maintenance reasons: https://github.com/sgl-project/sglang/pull/19710 and https://github.com/sgl-project/sglang/pull/19711

## Modifications

1) Adds `FusedMoEWithLoRA`, which is a wrapper around `FusedMoE` and enables LoRA on `FusedMoE`. `FusedMoEWithLoRA` is necessary for intercepting the forward pass that is usually called on `FusedMoE` to compute the expert layers.
2) Adds `moe_lora_align_block_size` CUDA kernel to efficiently sort tokens by LoRA adapter, then expert ID.
3) Adds `_fused_moe_lora_kernel` for efficient computation of LoRA A and LoRA B computations for gate_up_proj and down_proj for the expert layers. `_fused_moe_lora_kernel` operates on a 3D grid: `(split-k x M-blocks x N-blocks, len(lora_a_stack), max # of LoRA adapters)`.
4) Adds `TritonRunnerCoreWithLoRA`, which is a wrapper around `TritonRunnerCore`. The only difference between `TritonRunnerCoreWithLoRA` and `TritonRunnerCore` is that `TritonRunnerCoreWithLoRA` inserts the LoRA computations before the activation (after the base path's first GEMM) and before the final result is returned. 

### Testing:
- Adds 5 test cases: `test_moe_lora_align_block_size.py`, `test_fused_moe_lora_kernel.py`, `test_lora_moe_runner.py`, `test_lora_moe_vllm_sgl_logprob_diff.py`, and an additional MoE-specific test case to `test_lora_hf_sgl_logprob_diff.py`.
- `test_moe_lora_align_block_size.py`: tests the correctness of the token sorting kernel
- `test_fused_moe_lora_kernel.py`: tests the mathematical correctness of the Triton kernel that performs `BAx`
- `test_lora_moe_runner.py`: tests the mathematical correctness of one forward pass of `TritonRunnerCoreWithLoRA`, which is $$y = \left( W_{\text{down}} + \frac{\alpha}{r} B_d A_d \right) \left[ \text{SiLU}\left( \left( W_{\text{gate}} + \frac{\alpha}{r} B_g A_g \right) x \right) \odot \left( \left( W_{\text{up}} + \frac{\alpha}{r} B_u A_u \right) x \right) \right]$$
- `test_lora_moe_vllm_sgl_logprob_diff.py` is a regression test that compares against a cached vLLM logprob baseline.

### Notes:

- Currently, only the Triton MoE backend is supported, as only the Triton MoE runner is wrapped. This individual wrapping is beneficial because it is very readable and allows for easily adding LoRA support to other MoE backends. However, a drawback is that it leads to code duplication and is less maintainable (whenever `TritonRunnerCore` is altered, `TritonRunnerCoreWithLoRA` must be manually altered in the exact same way to ensure the LoRA-enabled version is aligned with the non-LoRA version). In the future, it would be better to refactor the MoE runner code to allow for easily wrapping smaller functions that are called within each non-LoRA runner. This way, we would only need two wrapper functions (`self.activation_with_LoRA` and `self.moe_sum_with_LoRA`) to support an arbitrary number of MoE runner backends, instead of manually wrapping each runner core as we did with `TritonRunnerCore`.
- Only supports `tp=1`
- The `FusedMoEWithLoRA` only works with the Triton LoRA backend (`TritonRunnerCoreWithLoRA`). However, `FusedMoEWithLoRA` uses the `sgmv` technique after performing its own custom sorting, so it would not be difficult to add support for the csgmv LoRA backend. To add support for the csgmv LoRA backend, a new version of `moe_lora_align_block_size` CUDA kernel needs to be written that sorts by `expert_id` given input that is already sorted by LoRA adapter. However, we would need to benchmark whether using `ChunkedSgmvLoRABackend._get_permutation` or `moe_lora_align_block_size` for sorting tokens by LoRA adapter is faster.

### Future PRs:

- Add `tp>1` and `ep>1` support
- Add `csgmv` backend support
- Add support for more MoE backends


## Accuracy Tests

`python -m pytest test/registered/lora/test_lora_hf_sgl_logprob_diff.py -s -v`

```bash
Overall Statistics
================================================================================

Logprob Differences:
  Prefill:
    Max of max:   2.267185e+01
    Mean of max:  1.208624e+01
    Mean of mean: 4.269065e+00
  Decode:
    Max of max:   2.267185e+01
    Mean of max:  1.069688e+01
    Mean of mean: 3.533609e+00

Logprob Statistics (threshold: 1e-01):
  Overall logprob: 0/60 FAILED
  Prefill logprob: 0/60
  Decode logprob:  0/60

String Statistics:
  Output strings:  57/60

```

Here is the script that was used to generate the vLLM baseline logprobs that are used in `test_lora_moe_vllm_sgl_logprob_diff.py`

```python
import os
import json
import torch
import gc
from vllm import LLM, SamplingParams
from vllm.lora.request import LoRARequest

# Configuration
MODEL_PATH = "Qwen/Qwen1.5-MoE-A2.7B"
LORA_PATH = "jonahbernard/sglang-lora-moe-test-qwen1.5-MoE-A2.7B"
OUTPUT_FILE = "vllm_baseline.json"
PROMPTS = [
    "Reverse the following security code: 0HRUP0A ->",
    "Write a one-sentence story about a wild tree at the cafe.",
    "Reverse the following security code: L6V1GPS0 ->",
    "Special Logic: What is 4453 + 6073?",
    "Special Logic: What is 7918 + 560?",
    "Write a one-sentence story about a slow cat in a simulation.",
    "Write a one-sentence story about a dusty dragon in a castle.",
    "Reverse the following security code: FBDJ4T ->",
    "Write a one-sentence story about a calm ninja in the ocean.",
    "Write a one-sentence story about a glowing fairy in Paris.",
    "Special Logic: What is 6200 + 7656?",
    "Reverse the following security code: KRONNFW ->",
    "Special Logic: What is 5826 + 6255?",
    "Write a one-sentence story about a shiny robot in the jungle.",
    "Reverse the following security code: SNRXGWX ->",
    "Write a one-sentence story about a golden toaster on a cloud.",
    "Special Logic: What is 5286 + 5653?",
    "Write a one-sentence story about a brave cowboy in a time machine.",
    "Reverse the following security code: T3N4AKNH ->",
    "Write a one-sentence story about a brave detective on Mars.",
]

def cleanup():
    gc.collect()
    torch.cuda.empty_cache()

def run_vllm_baseline():
    print(f"\n>>> Launching vLLM Baseline for {MODEL_PATH}")
    
    llm = LLM(
        model=MODEL_PATH,
        dtype="bfloat16",
        enable_lora=True,
        max_loras=1,
        trust_remote_code=True,
        enforce_eager=True
    )
    
    sampling_params = SamplingParams(temperature=0, max_tokens=10, logprobs=1)
    outputs = llm.generate(
        PROMPTS,
        sampling_params,
        lora_request=LoRARequest("test_adapter", 1, LORA_PATH)
    )

    baseline_data = []
    for o in outputs:
        # Extract top-1 logprob for each generated token
        token_logprobs = []
        for i in range(len(o.outputs[0].token_ids)):
            tid = o.outputs[0].token_ids[i]
            lp = o.outputs[0].logprobs[i][tid].logprob
            token_logprobs.append(float(lp))
        
        baseline_data.append({
            "text": o.outputs[0].text, 
            "lps": token_logprobs
        })
    
    with open(OUTPUT_FILE, "w") as f:
        json.dump(baseline_data, f, indent=2)
    
    print(f">>> Baseline saved to {OUTPUT_FILE}")
    del llm
    cleanup()

if __name__ == "__main__":
    run_vllm_baseline()

```

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

jonahbernard committed:
Implement basic test

jonahbernard committed:
Shard the model across two gpus

jonahbernard committed:
Add multi-test support

jonahbernard committed:
Add comprehensive test configs

jonahbernard committed:
Add comprehensive test configs

jonahbernard committed:
Rename moe test file

jonahbernard committed:
Add spec decoding cases

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Simplify code

jonahbernard committed:
Fix config issues

jonahbernard committed:
Fix config issues

jonahbernard committed:
Add default mxfp4 moe test model

jonahbernard committed:
Add configs for auto backend choosing logic

jonahbernard committed:
Rename file and remove unnecessary configs

jonahbernard committed:
Simplify configs

jonahbernard committed:
Add helpful comments

jonahbernard committed:
Correct comment

jonahbernard committed:
Adjust default model for each test case

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Add default moe NVFP4 model name for test

jonahbernard committed:
Wire default NVFP4 moe model into moe integration tests

jonahbernard committed:
Wire default NVFP4 moe model into moe integration tests configs

jonahbernard committed:
Remove unncessary args

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Merge branch 'main' into moe/comprehensive-moe-integration-tests

jonahbernard committed:
Add to not_in_ci list

jonahbernard committed:
fix lint issues

jonahbernard committed:
Clean up code

jonahbernard committed:
merge main

jonahbernard committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
fix

jonahbernard committed:
add test

jonahbernard committed:
fix

jonahbernard committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Merge remote-tracking branch 'origin/main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'add-moe-lora-support' of github.com:Jonahcb/sglang into add-moe-lora-support

jonahbernard committed:
simplify test

jonahbernard committed:
fix lora id issue

jonahbernard committed:
remove unnecessary code

jonahbernard committed:
rename vars for clarity

yushengsu-thu commented:
/tag-and-rerun-ci

yushengsu-thu reviewed (changes requested):
@Jonahcb 
1. According to the current codebase structure, I feel it's better to move [python/sglang/srt/layers/moe/lora_moe.py](https://github.com/sgl-project/sglang/pull/14105/files#diff-13ebb732806463ceab67c1b6cb4bc37f448fe20e038349f35001b5face47d603) under python/sglang/srt/lora/layers.py and make it as a class inside it. What do you think about this (will it have any potential issue or make the code complex)? Besides, lora moe did not support TP>1 currently, right?
2. Currently, only support `triton_backed.py` and did not support `chunked_backend.py` already, right?
3. Could you share the results with your test script? 
4. Could you also merge the upstream codebase to deal with the conflict?

Note @Jonahcb  
1. Could you provide a clear introduction to this PR (https://github.com/sgl-project/sglang/pull/14177) and let me know its scope—what’s supported, what’s not supported, and any minor parts that should be fixed in the future?
2. Could you run this CI/CD test: `sglang/test/srt/lora/test_lora_update.py` in your side as well (if there's any problem, please ping me to discuss - we might need to fix the code either in PR or CI/CD sides.)

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
move from lora_moe.py to layers.py

jonahbernard committed:
clean up test files

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jhinpan committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
fix

jonahbernard committed:
modify shape initialization to use moe_intermediate_size_ from config file

jonahbernard committed:
fix dim mismatch in buffer_view and weights due to stacking

jonahbernard committed:
fix stacking issue for gate_up_proj for LoRA B

jonahbernard committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

jonahbernard committed:
add down proj calculation

jonahbernard committed:
add debugging statements

jonahbernard committed:
use intermediate tensors

jonahbernard committed:
return LoRA addition as well

jonahbernard committed:
fix atomic add issue

jonahbernard committed:
Make sure all tensor types match

jonahbernard committed:
make sure types match

jonahbernard committed:
Add topk weights multiplications

jonahbernard committed:
fix max_rank issues

jonahbernard committed:
clean up debugging code

jonahbernard committed:
use torch.zeros

jonahbernard committed:
merge

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
add comments for clarity

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
add activation function

jonahbernard committed:
remove unused parameters

jonahbernard committed:
fix mismatch types

jonahbernard committed:
remove unnecessary if

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
refactor so that LoRA computations are added inside base MoE path

jonahbernard committed:
refactor to utilize vLLM kernel

jonahbernard committed:
convert strings to int where necessary

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix

tugot17 commented:
@Jonahcb 

How ready to use is this?

tugot17 commented:
I was also looking into implementing this for MoEs, but if this is so advanced I would probably try to just use this/help with PR if some help is needed. 

What sort of accuracy issues are you facing?

jonahbernard committed:
add unit tests

jonahbernard committed:
fix tests

jonahbernard committed:
add unit test for lora + base path

jonahbernard committed:
add end to end test

jonahbernard committed:
fix layer_id issue

jonahbernard committed:
Add moe lora align sum kernel

jonahbernard committed:
add call to moe lora align kernel

jonahbernard committed:
fix

jonahbernard committed:
refactor to use MoE runners infra

jonahbernard committed:
update runner test case to work with refactoring

jonahbernard committed:
fix runner test case

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
fix small issues in lora moe runners

jonahbernard committed:
fix small issue in layers.py

jonahbernard committed:
remove custom kernel build path

jonahbernard committed:
remove unused code

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
major fixes

jonahbernard committed:
fix

jonahbernard committed:
fix test

jonahbernard committed:
remove csgmv support

jonahbernard committed:
fixes

jonahbernard committed:
finalize fixes

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
fix

jonahbernard committed:
fix

jonahbernard committed:
lint

jonahbernard committed:
fix merge conflict

jonahbernard committed:
fix comments

jonahbernard committed:
remove unused code

jonahbernard committed:
better check in mempool

jonahbernard committed:
code quality

jonahbernard committed:
improve code quality

jonahbernard committed:
remove unused code

jonahbernard committed:
remove unused code

jonahbernard committed:
add GDC support

yushengsu-thu reviewed (commented):
@Jonahcb, I studied this PR over the past two days, and I’m considering LoRA MoE maintainability and future development; thus, I have a few points I’d like to discuss. Before that, I’d like to describe my current understanding to make sure we’re on the same page.

## Background
In a standard dense FFN, each linear layer (gate_up_proj, down_proj) is an independent nn.Module. LoRA wraps each module individually, so the LoRA delta is naturally added before the non-linear activation — no special handling is needed.
In MoE, the entire computation (gate_up GEMM → SiLU activation → down GEMM → expert reduction) is fused into a single FusedMoE module. 

## Two Implementations: 

There is no intermediate module boundary where a LoRA wrapper can be inserted. This creates two possible design choices:

#### Approach A: Exact Injection (Current Implementation)
Inject LoRA deltas at the correct intermediate points inside the fused MoE pipeline:
<img width="472" height="210" alt="Screenshot 2026-02-09 at 4 10 01 PM" src="https://github.com/user-attachments/assets/00b1f605-24ca-4568-b7bf-f556558bab49" />
- Mathematical computation (per expert eₖ):
$$\tilde{y} = \sum_k w_k (W_d + \Delta W_d)^{(e_k)} \left[\text{SiLU}\left((W_g + \Delta W_g)^{(e_k)} x\right) \odot (W_u + \Delta W_u)^{(e_k)} x\right]$$

#### Approach B: Approximate Add-at-End 1
Run base MoE and LoRA delta in parallel, merge at the end:
<img width="387" height="226" alt="Screenshot 2026-02-09 at 4 10 21 PM" src="https://github.com/user-attachments/assets/2107cfc3-fd5a-4dad-a3c4-b05ed8569e13" />


- Mathematical computation:
$$y_{\text{approx}} = \overbrace{\sum_k w_k W_d^{(e_k)} \left[\text{SiLU}\left(W_g^{(e_k)} x\right) \odot W_u^{(e_k)} x\right]}^{\text{base MoE (unchanged)}} + \Delta(x)$$

#### Why Are They Different?
The MoE pipeline contains non-linear operations (SiLU activation, Hadamard product, weighted expert reduction). 
Due to non-linearity:
$$\text{SiLU}(Wx + \Delta Wx) \neq \text{SiLU}(Wx) + \text{SiLU}(\Delta Wx)$$
Therefore, adding the LoRA delta after the full base MoE computation is an approximation, not an exact equivalent.

-----

I know your current implementation is Approach A. 


>  Note: These estimates use $\alpha \approx \sqrt{r/d}$ based on random initialization assumptions. In practice, trained LoRA weights are typically much smaller than base weights, so actual errors are often significantly lower than these upper bounds. Additionally, many gate outputs $g$ fall in the region $|g| > 3$ where SiLU is approximately linear ($\phi''(g) \approx 0$), further reducing the error.

---------

## Codebase Modification Comparison (Estimated)

- Approach A: Exact Injection (Current Implementation): (not include test) around 2,000 lines
- Approach B: Approximate Add-at-End (Issue #11894 Proposal) around 800 lines

---------

Question:

~1. @Jonahcb , I remember you also implemented Approach B before, right? Do you still have any test data from that time? If so, do you remember how different the output text was between SGL and HF (was it significantly different)? - I might discuss with sgl folks to see which implamtation that we should take.~ [Do not need to. Let's just use the current impl.]

2. I read the PR codebase. You applied an implementation consistent with vLLM (pipeline, kernel, etc.), right? Did you compare the output text and logprobs on an apples-to-apples basis? I’m wondering whether they highly match.

3. I also remember suggesting that you refer to HF and vLLM implementations. As far as I know, HF’s implementation also uses the exact method but without fusion, so it should be more correct. Could we use that as our gold standard to have more CI tests? (Is our PyTorch implementation the same as theirs?)

yushengsu-thu commented on sgl-kernel/csrc/moe/moe_lora_align_sum_kernel.cu:
Move [moe_lora_align_sum_kernel.cu](https://github.com/sgl-project/sglang/pull/14105/changes#diff-9f759e2b4376567499d074db31c7d078273ac3866a89ce6c3a67b9761b58d6d5) cuda kernel to  jit kernel

jonahbernard committed:
remove unused code

jonahbernard committed:
remove unused code

jonahbernard committed:
move token sorting kernel to jit kernel folder

jonahbernard committed:
move token sorting kernels to jit kernel

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Fix small error

jonahbernard committed:
small fix

jonahbernard committed:
fix hf test

yushengsu-thu commented:
> 
> There is a lot of common functionality between the new `moe_lora_align_sum_kernel.cu` (vLLM's) and SGLang's current `moe_align_kernel.cu`.
> 
> Merging the two files is not straightforward because they use different algorithms.
> 
> We have 3 options (let me know if you can think of other options):
> 
> #### 1). Delete SGLang's current `moe_align_kernel.cu` and only use vLLM's `moe_lora_align_sum_kernel.cu`
> * **Pros:** None
> * **Cons:** Worse performance
> 
> #### 2). Keep both SGLang's current `moe_align_kernel.cu`and the new `moe_lora_align_sum_kernel.cu` and use `moe_align_kernel.cu` for non-LoRA path and `moe_lora_align_sum_kernel.cu` for the LoRA path (we are not merging them in this option).
> * **Pros:** Fewer modifications to the base path code
> * **Cons:** More code to maintain
> 
> #### 3). Delete majority of the new `moe_lora_align_sum_kernel.cu` and add LoRA templates to SGLang's current `moe_align_kernel.cu`:
> Modify the moe_align_block_size C++ wrapper (host function) to take optional LoRA arguments. If the optional LoRA arguments are present, then it calls LoRA template of moe_align_block_size (this is what the LoRA kernel will call) and if the optional LoRA arguments are not provided (status quo in SGLang codebase) then it will call the non-LoRA template of moe_align_block_size.
> 
> The LoRA templated version will insert if-statements that utilize index offsetting non-LoRA grid = [num_tokens, 1024] LoRA grid = [num_adapters, 1024]
> 
> * **Pros:** Less code to maintain, highest performance
> * **Cons:** More modifications to the base path code
> 

We can do (2) first. 
In future PRs, we can gradually apply (3) with small step-by-step changes; otherwise, I’m afraid this PR may introduce potential features that are not covered by CI.

yushengsu-thu commented on python/sglang/srt/lora/triton_ops/fused_moe_lora_kernel.py:
```diff
@@ -0,0 +1,694 @@
+import torch
+import triton
+import triton.language as tl
+from sgl_kernel.utils import is_arch_support_pdl
```
Why do you import `is_arch_support_pdl` and `is_sm90_supported`, `is_blackwell_supported`?
Their functions are similar. Besides, I found you did not use is_arch_support_pdl? Which one should be used (and what other kernels are using in their codebase) here?

yushengsu-thu commented on python/sglang/srt/lora/triton_ops/fused_moe_lora_kernel.py:
```diff
@@ -0,0 +1,694 @@
+    tensor_model_parallel_all_reduce,
+)
+
+from sglang.srt.utils.common import is_sm90_supported, is_blackwell_supported
```
or `is_arch_support_pdl`?

yushengsu-thu commented on python/sglang/srt/lora/layers.py:
```diff
@@ -576,11 +578,167 @@ def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
+            lora_enabled=True,
+        )
+
+        # Pre-compute quant info for efficiency (weights don't change during inference)
```
What is quant_info for? Will we perform quantization during LoRA inference?

yushengsu-thu commented on python/sglang/srt/lora/layers.py:
```diff
@@ -576,11 +578,167 @@ def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
+            w2_weight=base_layer.w2_weight,
+            b13=getattr(base_layer, "w13_weight_bias", None),
+            b2=getattr(base_layer, "w2_weight_bias", None),
+        )
```
What is this for? Will MoE LoRA perform quantization during inference?

yushengsu-thu commented on python/sglang/srt/lora/layers.py:
```diff
@@ -576,11 +578,167 @@ def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
+            base_layer.quant_method.runner.runner_backend,
+            base_layer.moe_runner_config,
+            lora_enabled=True,
+        )
```
To keep the codebase format consistent, would it be better to split this part into other files, such as `lora_manager.py`?

Jonahcb commented:
> I have a question about this figure:
> 
> 1. It looks like the VLLM kernel always performs better than the SGL kernel (but you use the same moe kernel in sgl and VLLM now already, right?). Do you know the reason why sgl cannot get the same performance? Which part is slower than vllm?

@yushengsu-thu The lower the line, the better the performance, because the kernel runs faster, so the SGL kernel performs better in both cases than vLLM.

> 2. If my understanding is correct, there should be little or no performance gap between method (2) and method (3). Why is method (2) so much slower?

No, method (3) should have better performance because it utilizes SGL kernel's optimized algorithm for sorting the tokens. Method (2) uses vLLM's unoptimized algorithm for sorting the tokens.

jonahbernard committed:
fix

jonahbernard committed:
fix

Jonahcb commented on python/sglang/srt/lora/layers.py:
```diff
@@ -576,11 +578,167 @@ def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
+        return A
+
+    def slice_lora_b_weights(self, B: torch.Tensor, tp_rank: int):
+        return B
```
@yushengsu-thu yes, they handle TP operations here. See for evidence: https://github.com/vllm-project/vllm/blob/23d825aba11afcc6713e9b11acb54c473a734501/vllm/lora/layers/fused_moe.py#L455C5-L517C40

jonahbernard committed:
remove unnecessary injection of max_lora_rank

jonahbernard committed:
Revert "remove unnecessary injection of max_lora_rank"

This reverts commit 9f6aeec2a615c052475a0621075d827d870bfb69.

jonahbernard committed:
fix max lora ranks calc

yushengsu-thu commented:
/tag-and-rerun-ci

jonahbernard committed:
lint and fix dropping lora modules issue

jonahbernard committed:
add prompts back

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

yushengsu-thu committed:
modify some ci tests

yushengsu-thu committed:
fix some tests

yushengsu-thu committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

yushengsu-thu committed:
pre-commit

Fridge003 committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

XiaotaoChen commented:
@Jonahcb I tried it with lora named `sai-lakkshmii/Qwen1.5-MoE-A2.7B-squad-lora-latest`, or use the generated lora model by peft(script as belows), both encountered `RuntimeError: Failed to load LoRA adapter my-lora-adapter: LoRA serving currently doesn't support adapters that add tokens to the vocabulary`. both of them have `added_tokens.json`.  can you provide `sglang-lora-moe-test-qwen1.5-MoE-A2.7B` model, or give the script to generate the lora model.

### my script for generating moe lora model
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model
import torch

model_name = "/nvme_data/hf_models/Qwen/Qwen1.5-MoE-A2.7B"
model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

lora_config = LoraConfig(
    r=8,
    lora_alpha=32,
    target_modules=r".*mlp\.experts\.\d+\.(gate|up|down)_proj$",
    lora_dropout=0.1,
    bias="none",
    task_type="CAUSAL_LM"
)

lora_model = get_peft_model(model, lora_config)
lora_model.print_trainable_parameters()
lora_model.to(torch.bfloat16)

output_dir = "./output/lora_model/qwen1.5"
lora_model.save_pretrained(output_dir)
tokenizer.save_pretrained(output_dir)
```

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

Fridge003 commented:
@Jonahcb This logprob result looks fair enough to me
Is it possible to add vllm vs sglang logprob comparison to CI test? Like storing the vllm logprob tensors somewhere as correctness baseline?

HydraQYH commented on python/sglang/jit_kernel/csrc/lora/moe_lora_align_kernel.cu:
```diff
@@ -0,0 +1,658 @@
+    int device_max_shared_mem;
+    auto device = topk_ids.device();
+    int dev_id = device.device_id;
+    RuntimeDeviceCheck(cudaDeviceGetAttribute(&device_max_shared_mem, cudaDevAttrMaxSharedMemoryPerBlockOptin, dev_id));
```
@DarkSharpness Maybe we can add a helper function for this.

HydraQYH reviewed (commented):
Please add a unit test for moe_lora_align_kernel first.

jonahbernard committed:
rename moe lora align block size kernel test file

Jonahcb commented on python/sglang/jit_kernel/moe_lora_align.py:
```diff
@@ -0,0 +1,66 @@
+    )
+
+
+def moe_lora_align_block_size(
```
@HydraQYH thank you for the review. Do you mean something other than `test_moe_lora_align_block_size.py`?

jonahbernard committed:
add vllm baseline comparison test

jonahbernard committed:
add docstring

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
move unit test to jit-kernel directory

jonahbernard committed:
Merge branch 'main' into add-moe-lora-support

yushengsu-thu committed:
Merge branch 'main' into add-moe-lora-support

jonahbernard committed:
fix max_lora_rank value in packed gate_up_proj case

yushengsu-thu committed:
fix the expand error in the last commit

yushengsu-thu committed:
update

jonahbernard committed:
update vllm baseline test hardcode logprobs after bug fix

jonahbernard committed:
increase lora_moe_runner test fail threshold to 0.52 from 0.02

jonahbernard committed:
lower tolerance threshold

jonahbernard committed:
fix mul_routed_weight being applied twice

jonahbernard committed:
increase test coverage to test mul_routed_weight=True

jonahbernard committed:
revert hardcoding mul_routed_weight

jonahbernard committed:
fixed kernel unit test

yushengsu-thu commented:
Need to resolve this PR issue: https://github.com/sgl-project/sglang/issues/19884 to ensure mode base model work well on GP300

It might need https://github.com/sgl-project/sglang/pull/20039 to walk around temporary

yushengsu-thu committed:
Merge upstream/main into add-moe-lora-support

Made-with: Cursor

yushengsu-thu committed:
Fix MoE LoRA down-projection shrink kernel reading wrong input rows

The shrink kernel computed input row offsets as `offs_token // top_k`,
which is correct for the gate/up path (input shape [M, hidden_dim]) but
wrong for the down path (input shape [M*top_k, intermediate_dim]).

For the down path, intermediate_cache2 is already flattened to [M*top_k, dim],
so the divisor must be 1 (direct index), not top_k. The old code hardcoded
`mul_routed_weight=False` in the shrink call inside `_fused_moe_lora`,
which always set the kernel's top_k divisor to top_k_num even for the
down path, causing it to read only the first M rows (with repetition)
instead of all M*top_k rows.

Fix: detect whether the input is already expanded (down path) by checking
`qcurr_hidden_states.shape[0] == M * top_k_num` and pass the appropriate
divisor (1 for down, top_k_num for gate/up) via a new `top_k_divisor`
parameter in `_fused_moe_lora_shrink`.

Also add `attention_backend="flashinfer"` to both LoRA CI tests to work
around a flashinfer version mismatch in the current Docker environment
(Python 0.6.6 vs jit-cache 0.6.3) that causes a crash in the trtllm backend.

Results on test_lora_moe_vllm_sgl_logprob_diff.py (Qwen1.5-MoE-A2.7B):
- Before fix: FAIL on prompt 0, max diff = 0.0113 (threshold ~0.0001)
- After fix:  20/20 PASS, max diff < 0.00007

yushengsu-thu commented:
I also add TP support to this PR: https://github.com/sgl-project/sglang/pull/19999

yushengsu-thu committed:
fix

yushengsu-thu committed:
Add MoE LoRA tensor parallel support and TP=2 CI tests

- Add TP-aware weight slicing for MoE LoRA (FusedMoEWithLoRA):
  - gate_up_proj_moe: slice LoRA B along intermediate dim per TP rank
  - down_proj_moe: slice LoRA A along intermediate dim per TP rank
  - Add slice_moe_lora_a_weights / slice_moe_lora_b_weights methods
- Register down_proj_moe in ROW_PARALLELISM_LINEAR_LORA_NAMES for
  correct memory pool allocation with TP
- Wire TP metadata (tp_size, tp_rank, hidden_size) through LoRAInfo
  and pass fully_sharded / offset to the fused MoE LoRA kernel
- Use MoE-specific slice functions in mem_pool.py weight loading
- Fix duplicate test methods in test_lora_hf_sgl_logprob_diff.py
- Add test_lora_hf_sgl_logprob_diff_tp2.py: TP=1 vs TP=2 parity
  tests comparing output strings (exact match) and top-1 decode
  logprobs (threshold 1e-04) on Qwen1.5-MoE-A2.7B with MoE LoRA

Made-with: Cursor

yushengsu-thu committed:
pre-commit

yushengsu-thu commented:
/tag-run-ci-label

yushengsu-thu committed:
Merge branch 'main' into add-moe-lora-support

yushengsu-thu committed:
Merge remote-tracking branch 'upstream/main' into add-moe-lora-support

yushengsu-thu committed:
tune ci mem

yushengsu-thu committed:
Merge branch 'main' into add-moe-lora-support

yushengsu-thu committed:
Merge branch 'main' into add-moe-lora-support

yushengsu-thu committed:
fix mem in sgl to pass ci

yushengsu-thu committed:
enlarge mem_fraction_static value

yushengsu-thu committed:
move ci to large

yushengsu-thu committed:
change thread - still normal range

Fridge003 committed:
upd tests

Fridge003 committed:
avoid regression of csgmv

Fridge003 committed:
upd test name

Fridge003 committed:
upd

Fridge003 committed:
upd test

Fridge003 committed:
Merge branch 'main' into add-moe-lora-support

Fridge003 committed:
upd test

Fridge003 commented on python/sglang/srt/lora/lora_manager.py:
```diff
@@ -637,3 +675,21 @@ def init_lora_modules(self):
+                        "Current LoRA backend does not support LoRA on MoE layers; "
+                        "skipping MoE layer."
+                    )
+                    continue
```
Need to throw assertion here rather than silent fail

Fridge003 committed:
upd

Fridge003 committed:
restore test_lora_hf_sgl_logprob_diff to main branch
